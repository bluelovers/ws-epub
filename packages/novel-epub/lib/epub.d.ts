import { IForeachArrayDeepReturn, IReturnRow } from 'node-novel-globby';
import EpubMaker from 'epub-maker2';
import { EnumEpubTypeName } from 'epub-maker2/src/epub-types';
import { INovelEpubReturnInfo } from './txt2epub3';
import { IInternalProcessEpubOptions, IInternalProcessVolumeOptions, IResolvableBluebird } from './types';
import { handleMarkdown } from './md';
import Bluebird = require('bluebird');
export declare const SymCache: unique symbol;
export declare const enum EnumPrefixIDType {
    VOLUME = "volume",
    CHAPTER = "chapter",
    IMAGE = "image",
    CONTRIBUTE = "contribute",
    FOREWORD = "foreword"
}
export declare const enum EnumPrefixIDTitle {
    IMAGE = "\u63D2\u5716",
    CONTRIBUTE = "CONTRIBUTE",
    FOREWORD = "FOREWORD"
}
export interface IEpubRuntimeReturnCacheVolumeRow {
    vol_key: string;
    dirname: string;
    value: IReturnRow;
}
export declare type IEpubRuntimeReturn = IForeachArrayDeepReturn<IReturnRow, any, {
    stat: INovelEpubReturnInfo["stat"];
}, {
    cache_vol: {
        [vol: string]: number;
    };
    cache_volume_row: IEpubRuntimeReturnCacheVolumeRow[];
    cache_top_subs: {
        [k: string]: Omit<IEpubRuntimeReturnCacheVolumeRow, 'value'>[];
    };
    prev_volume_row?: IEpubRuntimeReturnCacheVolumeRow;
    prev_volume_title: string;
    prev_volume_dir: string;
    count_idx: number;
    count_f: number;
    count_d: number;
    cacheTreeSection: Record<string, IEpubMakerSectionWithCache>;
    prev_volume?: IEpubMakerSectionWithCache;
    _new_top_level?: IEpubMakerSectionWithCache;
    _old_top_level?: IEpubMakerSectionWithCache;
}>;
export interface IEpubMakerSectionWithCache extends EpubMaker.Section {
    [SymCache]?: {
        cover?: boolean;
        image?: boolean;
        contribute?: boolean;
        foreword?: boolean;
    };
}
export declare function _handleVolume(volume: IEpubMakerSectionWithCache, dirname: string, _data_: IInternalProcessVolumeOptions): Bluebird<void>;
export declare function addMarkdown(options: {
    volume: IEpubMakerSectionWithCache;
    dirname: string;
    _data_: IInternalProcessVolumeOptions;
    file: string;
    epubType: EnumEpubTypeName;
    epubPrefix: EnumPrefixIDType;
    epubTitle: string | EnumPrefixIDTitle;
}): Bluebird<EpubMaker.Section>;
export declare function makePrefixID(count_idx: number, prefix: EnumPrefixIDType): string;
export declare function makeVolumeID(count_idx: number): string;
export declare function makeChapterID(count_idx: number): string;
export interface IAttachMetaData {
    images: Record<string, string>;
}
export declare function getAttachMeta(dirname: string): Promise<IAttachMetaData>;
export declare function getAttachMetaByRow(row: IReturnRow): Promise<IAttachMetaData>;
export declare function _handleVolumeImage(volume: IEpubMakerSectionWithCache | EpubMaker, dirname: string, _data_: IInternalProcessVolumeOptions): Bluebird<string[]>;
export declare function _handleVolumeImageEach(ls: Omit<IEpubRuntimeReturnCacheVolumeRow, 'value'>[], _data_: IInternalProcessVolumeOptions): Bluebird<string[][]>;
export declare function _hookAfterVolume(ls: IEpubRuntimeReturn["temp"]["cache_volume_row"], _data_: IInternalProcessVolumeOptions, afterVolumeTasks: ((volume: IEpubMakerSectionWithCache, dirname: string, _data_: IInternalProcessVolumeOptions, row: IEpubRuntimeReturnCacheVolumeRow) => IResolvableBluebird<unknown>)[]): Bluebird<{
    index: number;
    row: IEpubRuntimeReturnCacheVolumeRow;
    volume: IEpubMakerSectionWithCache;
    mapData: {
        index: number;
        fn: (volume: IEpubMakerSectionWithCache, dirname: string, _data_: IInternalProcessVolumeOptions, row: IEpubRuntimeReturnCacheVolumeRow) => IResolvableBluebird<unknown>;
        ret: unknown;
    }[];
}[]>;
export declare function _hookAfterEpub(epub: EpubMaker, _data_: IInternalProcessEpubOptions, afterEpubTasks: ((epub: EpubMaker, _data_: IInternalProcessEpubOptions) => IResolvableBluebird<unknown>)[]): Bluebird<{
    epub: EpubMaker;
    mapData: {
        index: number;
        fn: (epub: EpubMaker, _data_: IInternalProcessEpubOptions) => IResolvableBluebird<unknown>;
        ret: unknown;
    }[];
}>;
export declare function addContributeSection(volume: IEpubMakerSectionWithCache, dirname: string, _data_: IInternalProcessVolumeOptions, row: IEpubRuntimeReturnCacheVolumeRow): Bluebird<boolean>;
export declare function createContributeSection(options: {
    target: IEpubMakerSectionWithCache | EpubMaker;
    mdReturn: ReturnType<typeof handleMarkdown>;
    processReturn: Partial<IEpubRuntimeReturn>;
}): EpubMaker.Section;
export declare function createMarkdownSection(options: {
    target: IEpubMakerSectionWithCache | EpubMaker;
    mdReturn: ReturnType<typeof handleMarkdown>;
    processReturn: Partial<IEpubRuntimeReturn>;
    epubType: EnumEpubTypeName;
    epubPrefix: EnumPrefixIDType;
    epubTitle: string | EnumPrefixIDTitle;
}): EpubMaker.Section;
export declare function _withSection(target: IEpubMakerSectionWithCache | EpubMaker, chapter: EpubMaker.Section): EpubMaker.Section;
