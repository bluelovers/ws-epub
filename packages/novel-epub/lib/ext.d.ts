export declare const allowExtImage: readonly ["apng", "bmp", "bpg", "gif", "heic", "heif", "jpeg", "jpg", "png", "svg", "webp", "ico", "jfif"];
export declare function toGlobExtImage(): readonly [`*.{${string}}`, `image/*.{${string}}`, `images/*.{${string}}`, `img/*.{${string}}`, `imgs/*.{${string}}`];
export declare function isAllowExtImage(ext: string): boolean;
