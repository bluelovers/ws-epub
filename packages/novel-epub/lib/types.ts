import { IOptions } from './txt2epub3';
import { EpubMaker } from 'epub-maker2';
import { EpubStore } from './store';
import { IAttachMetaData, IEpubRuntimeReturn } from './epub';
import MarkdownIt = require('markdown-it');
import { ITSResolvable, ITSPartialWith, ITSUnpackedPromiseLike, ITSRequiredWith } from 'ts-type';
import Bluebird = require('bluebird');

export interface IInternalProcessOptions
{
	store: EpubStore,
	epub: EpubMaker,
	epubOptions: IOptions,
	cwd: string,
	cwdRoot: string,
}

export interface IInternalProcessContextOptions extends IInternalProcessOptions
{
	vid: string,
	attach: IAttachMetaData,
}

export interface IMdEnv
{
	title?: string,
	[k: string]: unknown,
}

export interface IInternalProcessMarkdownItOptions extends Partial<IInternalProcessContextOptions>
{
	mdOptions?: MarkdownIt.Options,
	md?: MarkdownIt,
	mdEnv?: IMdEnv,
}

export interface IInternalProcessVolumeOptions extends Omit<IInternalProcessOptions, 'cwd'>
{
	processReturn: Partial<IEpubRuntimeReturn>,
	cwd?: string,
}

export interface IInternalProcessEpubOptions extends IInternalProcessOptions
{
	processReturn: IEpubRuntimeReturn,
}

export type IResolvableBluebird<T> = ITSResolvable<T> | Bluebird<T>

