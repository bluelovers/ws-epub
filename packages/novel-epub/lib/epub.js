"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._withSection = exports.createMarkdownSection = exports.createContributeSection = exports.addContributeSection = exports._hookAfterEpub = exports._hookAfterVolume = exports._handleVolumeImageEach = exports._handleVolumeImage = exports.getAttachMetaByRow = exports.getAttachMeta = exports.makeChapterID = exports.makeVolumeID = exports.makePrefixID = exports.addMarkdown = exports._handleVolume = exports.EnumPrefixIDTitle = exports.EnumPrefixIDType = exports.SymCache = void 0;
const epub_maker2_1 = require("epub-maker2");
const log_1 = require("./log");
const util_1 = require("./util");
const util_2 = require("epub-maker2/src/lib/util");
const crlf_normalize_1 = require("crlf-normalize");
const array_hyper_unique_1 = require("array-hyper-unique");
const store_1 = require("./store");
const html_1 = require("./html");
const md_1 = require("./md");
const ext_1 = require("./ext");
const str_util_1 = require("str-util");
const Bluebird = require("bluebird");
const path = require("upath2");
const fs = require("fs-iconv");
const novelGlobby = require("node-novel-globby/g");
exports.SymCache = Symbol('cache');
var EnumPrefixIDType;
(function (EnumPrefixIDType) {
    EnumPrefixIDType["VOLUME"] = "volume";
    EnumPrefixIDType["CHAPTER"] = "chapter";
    EnumPrefixIDType["IMAGE"] = "image";
    EnumPrefixIDType["CONTRIBUTE"] = "contribute";
    EnumPrefixIDType["FOREWORD"] = "foreword";
})(EnumPrefixIDType = exports.EnumPrefixIDType || (exports.EnumPrefixIDType = {}));
var EnumPrefixIDTitle;
(function (EnumPrefixIDTitle) {
    EnumPrefixIDTitle["IMAGE"] = "\u63D2\u5716";
    EnumPrefixIDTitle["CONTRIBUTE"] = "CONTRIBUTE";
    EnumPrefixIDTitle["FOREWORD"] = "FOREWORD";
})(EnumPrefixIDTitle = exports.EnumPrefixIDTitle || (exports.EnumPrefixIDTitle = {}));
function _handleVolume(volume, dirname, _data_) {
    return Bluebird
        .resolve(null)
        .then(async function () {
        const { processReturn, epub } = _data_;
        const { stat } = processReturn.data;
        let vid = volume.id;
        if (!volume[exports.SymCache].cover) {
            volume[exports.SymCache].cover = true;
            let file = path.join(dirname, 'README.md');
            let meta = await util_1.fsLowCheckLevelMdconfAsync(file).catch(e => null);
            //console.log(file, meta);
            await Bluebird.resolve(novelGlobby.globby([
                'cover.*',
            ], {
                cwd: dirname,
                absolute: true,
            }))
                .then(async (ls) => {
                if (ls.length) {
                    let ext = path.extname(ls[0]);
                    let name = `${vid}-cover${ext}`;
                    epub.withAdditionalFile(ls[0], null, name);
                    return name;
                }
                else if (fs.existsSync(file)) {
                    if (meta && meta.novel) {
                        if (meta.novel.cover) {
                            let ext = '.png';
                            let basename = `${vid}-cover`;
                            let name = `${basename}${ext}`;
                            let data = typeof meta.novel.cover === 'string' ? {
                                url: meta.novel.cover,
                            } : meta.novel.cover;
                            data.ext = null;
                            data.basename = basename;
                            epub.withAdditionalFile(data, null, name);
                            return name;
                        }
                    }
                }
            })
                .then(function (name) {
                let _ok = false;
                let data = {};
                if (name) {
                    _ok = true;
                    data.cover = {
                        name,
                    };
                    stat.image += 1;
                }
                if (meta && meta.novel) {
                    if (meta.novel.preface) {
                        _ok = true;
                        //data.content = crlf(meta.novel.preface);
                        data.content = util_2.htmlPreface({
                            infoPreface: meta.novel.preface,
                        }).infoPrefaceHTML;
                    }
                }
                //console.log(name, _ok);
                if (_ok) {
                    return data;
                }
                return null;
            })
                .tap(function (data) {
                if (data) {
                    //console.log(volume);
                    volume.setContent(data, true);
                }
            });
        }
        if (!volume[exports.SymCache].foreword) {
            volume[exports.SymCache].foreword = true;
            let file = path.join(dirname, 'FOREWORD.md');
            if (fs.pathExistsSync(file)) {
                addMarkdown({
                    volume,
                    dirname,
                    _data_,
                    file,
                    epubType: "foreword" /* FOREWORD */,
                    epubPrefix: "foreword" /* FOREWORD */,
                    epubTitle: "FOREWORD" /* FOREWORD */,
                });
            }
        }
    });
}
exports._handleVolume = _handleVolume;
function addMarkdown(options) {
    return Bluebird.resolve(options)
        .then(async () => {
        let { volume, _data_, file, dirname } = options;
        const vid = volume.id;
        const { processReturn } = _data_;
        let source = await fs.readFile(file);
        let mdReturn = md_1.handleMarkdown(source, {
            ..._data_,
            cwd: dirname,
            vid,
        });
        let chapter = createMarkdownSection({
            target: volume,
            mdReturn,
            processReturn,
            epubType: options.epubType,
            epubTitle: options.epubTitle,
            epubPrefix: options.epubPrefix,
        });
        return chapter;
    });
}
exports.addMarkdown = addMarkdown;
function makePrefixID(count_idx, prefix) {
    return `${prefix}${(count_idx).toString().padStart(6, '0')}`;
}
exports.makePrefixID = makePrefixID;
function makeVolumeID(count_idx) {
    return makePrefixID(count_idx, "volume" /* VOLUME */);
}
exports.makeVolumeID = makeVolumeID;
function makeChapterID(count_idx) {
    return makePrefixID(count_idx, "chapter" /* CHAPTER */);
}
exports.makeChapterID = makeChapterID;
function getAttachMeta(dirname) {
    return util_1.fsLowCheckLevelMdconfAsync(path.join(dirname, 'ATTACH.md'))
        // @ts-ignore
        .then((v) => {
        if (v.attach) {
            return v.attach;
        }
        // @ts-ignore
        else if (v.novel && v.novel.attach) {
            // @ts-ignore
            return v.novel.attach;
        }
        return v;
    })
        // @ts-ignore
        .then((attach) => {
        if (attach && attach.images) {
            attach.images = Object.entries(attach.images)
                .reduce((a, [k, v]) => {
                a[k] = v;
                let k2 = k.toString().toLowerCase();
                if (a[k2] == null) {
                    a[k2] = v;
                }
                k2 = k2.toUpperCase();
                if (a[k2] == null) {
                    a[k2] = v;
                }
                return a;
            }, {});
        }
        return attach;
    })
        .catch(e => null);
}
exports.getAttachMeta = getAttachMeta;
const AttachMetaMap = new Map();
async function getAttachMetaByRow(row) {
    if (!AttachMetaMap.has(row.path_dir)) {
        let data = await getAttachMeta(row.path_dir);
        AttachMetaMap.set(row.path_dir, data);
    }
    return AttachMetaMap.get(row.path_dir);
}
exports.getAttachMetaByRow = getAttachMetaByRow;
function _handleVolumeImage(volume, dirname, _data_) {
    const globImages = [
        ...ext_1.toGlobExtImage(),
        '!cover.*',
        '!*.txt',
    ];
    const baseImagePath = 'image';
    const failbackExt = '.png';
    return Bluebird.resolve(null)
        .then(async function () {
        if (volume[exports.SymCache].image) {
            return [];
        }
        const { processReturn, epub, epubOptions, store, cwdRoot } = _data_;
        const { stat } = processReturn.data;
        let vid;
        if (volume instanceof epub_maker2_1.default) {
            vid = '';
        }
        else {
            vid = volume.id;
        }
        volume[exports.SymCache].image = true;
        return novelGlobby.globby(globImages, {
            cwd: dirname,
            absolute: true,
        })
            .then(async (ls) => {
            let arr = [];
            let arr2 = [];
            for (let i in ls) {
                let img = ls[i];
                /*
                let ext = path.extname(img);

                let basename = path.basename(img, ext);

                // @ts-ignore
                let name = slugify(basename);

                if (!name || arr.includes(name))
                {
                    name = hashSum([img, i, name]);
                }

                //name = `${vid}/${i}-` + name + ext;
                name = `${vid}/` + name + ext;

                arr.push('image/' + name);

                epub.withAdditionalFile(img, 'image', name);
                 */
                let ret = store_1.handleAttachFile(img, {
                    vid,
                    epub,
                    epubOptions,
                    store,
                    basePath: baseImagePath,
                    failbackExt,
                    cwd: dirname,
                    cwdRoot,
                });
                if (ret && !arr.includes(ret.returnPath)) {
                    arr.push(ret.returnPath);
                    arr2.push({
                        attr: ` alt="（IMG：${str_util_1.toFullWidth('' + ret.data.basename)}）"`,
                        src: ret.returnPath,
                    });
                }
            }
            let md_attach = await getAttachMeta(dirname);
            if (md_attach && (md_attach.images)) {
                Object.values(md_attach.images)
                    .forEach(v => {
                    if (v) {
                        let ret = store_1.handleAttachFile(v, {
                            vid,
                            epub,
                            epubOptions,
                            store,
                            basePath: baseImagePath,
                            failbackExt,
                            cwd: dirname,
                            cwdRoot,
                        });
                        if (ret && !arr.includes(ret.returnPath)) {
                            arr.push(ret.returnPath);
                            arr2.push({
                                attr: ` alt="（插圖${str_util_1.toFullWidth(v)}）"`,
                                src: ret.returnPath,
                            });
                        }
                    }
                });
            }
            arr = array_hyper_unique_1.array_unique(arr);
            if (arr.length) {
                if (volume instanceof epub_maker2_1.default.Section) {
                    if (volume.content && volume.content.cover && volume.content.cover.name) {
                        arr.unshift(volume.content.cover.name);
                    }
                }
                let chapter = new epub_maker2_1.default.Section("non-specific backmatter" /* NON_SPECIFIC_BACKMATTER */, makePrefixID(processReturn.temp.count_idx++, "image" /* IMAGE */), {
                    title: "\u63D2\u5716" /* IMAGE */,
                    content: arr2.reduce(function (a, b) {
                        let html = html_1.novelImage(b.src, {
                            attr: b.attr,
                        });
                        a.push(html);
                        return a;
                    }, []).join(crlf_normalize_1.LF),
                }, true, false);
                stat.image += arr.length;
                //volume.withSubSection(chapter);
                _withSection(volume, chapter);
            }
            return ls;
        });
    });
}
exports._handleVolumeImage = _handleVolumeImage;
function _handleVolumeImageEach(ls, _data_) {
    const { processReturn, epub, store, epubOptions, cwd } = _data_;
    const temp = processReturn.temp;
    return Bluebird
        .resolve(array_hyper_unique_1.array_unique(ls))
        .mapSeries(async function (row) {
        let key = row.vol_key;
        let volume = temp.cacheTreeSection[key];
        return _handleVolumeImage(volume, row.dirname, _data_)
            .tap(function (ls) {
            if (0 && ls.length) {
                log_1.console.log({
                    volume,
                    ls,
                });
            }
        });
    });
}
exports._handleVolumeImageEach = _handleVolumeImageEach;
function _hookAfterVolume(ls, _data_, afterVolumeTasks) {
    const { processReturn, epub, store, epubOptions, cwd } = _data_;
    const temp = processReturn.temp;
    ls = array_hyper_unique_1.array_unique(ls);
    return Bluebird
        .resolve(ls)
        .mapSeries(async function (row, index) {
        let key = row.vol_key;
        let volume = temp.cacheTreeSection[key];
        return Bluebird.props({
            index,
            row,
            volume,
            mapData: Bluebird.mapSeries(afterVolumeTasks, async (fn, index) => {
                return {
                    index,
                    fn,
                    ret: await fn(volume, row.dirname, _data_, row),
                };
            })
        });
    });
}
exports._hookAfterVolume = _hookAfterVolume;
function _hookAfterEpub(epub, _data_, afterEpubTasks) {
    return Bluebird
        .resolve(epub)
        .then(async function (epub) {
        return Bluebird.props({
            epub,
            mapData: Bluebird.mapSeries(afterEpubTasks, async (fn, index) => {
                return {
                    index,
                    fn,
                    ret: await fn(epub, _data_),
                };
            })
        });
    });
}
exports._hookAfterEpub = _hookAfterEpub;
function addContributeSection(volume, dirname, _data_, row) {
    return Bluebird.resolve(volume)
        .then(async (volume) => {
        if (volume[exports.SymCache].contribute != null) {
            return;
        }
        volume[exports.SymCache].contribute = false;
        return novelGlobby.globby([
            'CONTRIBUTE.md',
        ], {
            cwd: dirname,
            absolute: true,
            deep: 0,
        })
            .then(async (ls) => {
            if (ls.length) {
                const vid = volume.id;
                const attach = await getAttachMetaByRow(row.value);
                const { processReturn } = _data_;
                let file = ls[0];
                let source = await fs.readFile(file);
                let mdReturn = md_1.handleMarkdown(source, {
                    ..._data_,
                    cwd: dirname,
                    vid,
                    attach,
                });
                let chapter = createContributeSection({
                    target: volume,
                    mdReturn,
                    processReturn,
                });
                return volume[exports.SymCache].contribute = true;
            }
        });
    });
}
exports.addContributeSection = addContributeSection;
function createContributeSection(options) {
    return createMarkdownSection({
        ...options,
        epubType: "non-specific backmatter" /* NON_SPECIFIC_BACKMATTER */,
        epubTitle: "CONTRIBUTE" /* CONTRIBUTE */,
        epubPrefix: "contribute" /* CONTRIBUTE */,
    });
}
exports.createContributeSection = createContributeSection;
function createMarkdownSection(options) {
    const { target, mdReturn, processReturn, } = options;
    let title = mdReturn.mdEnv.title || options.epubTitle;
    let chapter = new epub_maker2_1.default.Section(options.epubType, makePrefixID(processReturn.temp.count_idx++, options.epubPrefix), {
        title,
        content: mdReturn.mdHtml,
    }, true, false);
    return _withSection(target, chapter);
}
exports.createMarkdownSection = createMarkdownSection;
function _withSection(target, chapter) {
    if (target instanceof epub_maker2_1.default) {
        target.withSection(chapter);
    }
    else {
        target.withSubSection(chapter);
    }
    return chapter;
}
exports._withSection = _withSection;
//# sourceMappingURL=epub.js.map