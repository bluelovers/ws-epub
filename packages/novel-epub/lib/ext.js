"use strict";
/**
 * Created by user on 2019/7/27.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAllowExtImage = exports.toGlobExtImage = exports.allowExtImage = void 0;
exports.allowExtImage = [
    'apng',
    'bmp',
    'bpg',
    'gif',
    'heic',
    'heif',
    'jpeg',
    'jpg',
    'png',
    'svg',
    'webp',
    'ico',
    'jfif',
];
function toGlobExtImage() {
    let exts = exports.allowExtImage.join(',');
    return [
        `*.{${exts}}`,
        `image/*.{${exts}}`,
        `images/*.{${exts}}`,
        `img/*.{${exts}}`,
        `imgs/*.{${exts}}`,
    ];
}
exports.toGlobExtImage = toGlobExtImage;
function isAllowExtImage(ext) {
    return exports.allowExtImage.includes(ext.replace(/^\./, '').toLowerCase());
}
exports.isAllowExtImage = isAllowExtImage;
//# sourceMappingURL=ext.js.map