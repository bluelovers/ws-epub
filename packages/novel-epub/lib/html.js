"use strict";
/**
 * Created by user on 2019/7/9.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitTxt = exports.novelImage = void 0;
const min_1 = require("cjk-conv/lib/zh/convert/min");
const tags_1 = require("./tags");
const store_1 = require("./store");
const str_util_1 = require("str-util");
const log_1 = require("./log");
const parse_txt_tag_1 = require("@node-novel/parse-txt-tag");
function novelImage(src, options = {}) {
    let { failback = '', attr = '' } = options || {};
    if (failback && failback.length) {
        failback = ` lowsrc="${failback}" `;
    }
    else {
        failback = '';
    }
    return `<figure class="fullpage ImageContainer page-break-before duokan-image-single"><div><img src="${src}" class="inner-image" ${failback} ${attr}/></div></figure>`;
}
exports.novelImage = novelImage;
function splitTxt(txt, plusData) {
    const { attach = {}, store, vid, epub, epubOptions, cwd } = plusData || {};
    const { images } = attach || {};
    if (epubOptions.iconv) {
        if (epubOptions.iconv === 'cn') {
            txt = min_1.tw2cn_min(txt);
        }
        else if (epubOptions.iconv === 'tw') {
            txt = min_1.cn2tw_min(txt);
        }
    }
    let context = txt
        .toString()
        .replace(/\r\n|\r(?!\n)|\n/g, "\n")
        .replace(/\u003C/g, '&lt;')
        .replace(/\u003E/g, '&gt;')
        .replace(/&lt;(img[^\n]+?)\/?&gt;/gm, function (...m) {
        //console.log(m);
        return `<${m[1].replace(/\/+$/, '')} class="inner-image"/>`;
    });
    context = parse_txt_tag_1.parse(context, {
        on: {
            img({ tagName, innerContext, }) {
                if (images && store && innerContext) {
                    let { id, input } = store_1.getAttachID(innerContext, {
                        images,
                    });
                    if (input) {
                        let ret = store_1.handleAttachFile(input, {
                            ...plusData,
                            store,
                            epubOptions,
                            epub,
                            vid,
                            failbackExt: '.jpg',
                            basePath: 'image',
                            cwd,
                        });
                        if (ret) {
                            let _options = {
                                attr: ` alt="（插圖${str_util_1.toFullWidth(id)}）"`,
                            };
                            if (ret.ok && !ret.isFile) {
                                _options.failback = ret.input;
                                return novelImage(ret.returnPath, _options);
                            }
                            return novelImage(ret.returnPath, _options);
                        }
                    }
                    else {
                        log_1.console.warn(`ID( ${innerContext} )，不存在於附件表內`);
                    }
                }
                else {
                    log_1.console.warn(`偵測到 ID( ${innerContext} )，但附件表不存在`);
                }
                return null;
            },
            default({ tagName, attr, innerContext, cache, attach, }) {
                if (tagName !== 'img' && !tags_1.allowedHtmlTagList.includes(tagName)) {
                    log_1.console.warn(`not support ${tagName}`, attr, innerContext);
                }
                return null;
            },
        }
    }).context;
    context = ('<div>' + context
        .replace(/^[ ]*[－＝\-—\=─–]{3,}[ ]*$/mg, '<hr/>')
        //.replace(/^([－＝\-—\=─═─＝=══－\-─—◆◇]+)$/mg, '<span class="overflow-line">$1</span>')
        .replace(/\n/g, '</div>\n<div>')
        + '</div>')
        .replace(/<div><hr\/><\/div>/g, '<hr class="linehr"/>')
        .replace(/<div>[ ]*([－＝—=─═─＝=══－\-─—～◆◇\*＊\+＊＊↣◇◆☆★■□☆◊▃\p{Punctuation}]+)[ ]*<\/div>/ug, '<div class="linegroup calibre1 overflow-line">$1</div>')
        .replace(/<div><\/div>/g, '<div class="linegroup softbreak">　 </div>')
        .replace(/<div>/g, '<div class="linegroup calibre1">');
    return context;
}
exports.splitTxt = splitTxt;
exports.default = splitTxt;
//# sourceMappingURL=html.js.map