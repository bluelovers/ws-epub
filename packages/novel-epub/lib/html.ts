/**
 * Created by user on 2019/7/9.
 */

import { IInternalProcessContextOptions } from './types';
import { IAttachMetaData } from './epub';
import { cn2tw_min, tw2cn_min } from 'cjk-conv/lib/zh/convert/min';
import { _fixRubyInnerContext, allowedHtmlTagList, reTxtHtmlTag, reTxtImgTag } from './tags';
import { getAttachID, handleAttachFile } from './store';
import { toHalfWidth, toFullWidth } from 'str-util';
import { console } from './log';
import { parse as parseNodeNovelTxtTag } from '@node-novel/parse-txt-tag';

export function novelImage(src: string, options: {
	failback?: string,
	attr?: string,
} = {})
{
	let { failback = '', attr = '' } = options || {};

	if (failback && failback.length)
	{
		failback = ` lowsrc="${failback}" `;
	}
	else
	{
		failback = '';
	}

	return `<figure class="fullpage ImageContainer page-break-before duokan-image-single"><div><img src="${src}" class="inner-image" ${failback} ${attr}/></div></figure>`;
}

export function splitTxt(txt, plusData?: IInternalProcessContextOptions)
{
	const { attach = {} as IAttachMetaData, store, vid, epub, epubOptions, cwd } = plusData || {};
	const { images } = attach || {} as IAttachMetaData;

	if (epubOptions.iconv)
	{
		if (epubOptions.iconv === 'cn')
		{
			txt = tw2cn_min(txt)
		}
		else if (epubOptions.iconv === 'tw')
		{
			txt = cn2tw_min(txt)
		}
	}

	let context: string = txt
		.toString()
		.replace(/\r\n|\r(?!\n)|\n/g, "\n")

		.replace(/\u003C/g, '&lt;')
		.replace(/\u003E/g, '&gt;')

		.replace(/&lt;(img[^\n]+?)\/?&gt;/gm, function (...m)
		{
			//console.log(m);

			return `<${m[1].replace(/\/+$/, '')} class="inner-image"/>`;
		})
	;

	context = parseNodeNovelTxtTag(context, {
		on:{
			img({
				tagName,
				innerContext,
			})
			{
				if (images && store && innerContext)
				{
					let { id, input } = getAttachID(innerContext, {
						images,
					});

					if (input)
					{
						let ret = handleAttachFile(input, {
							...plusData,
							store,
							epubOptions,
							epub,
							vid,
							failbackExt: '.jpg',
							basePath: 'image',
							cwd,
						});

						if (ret)
						{
							let _options: Parameters<typeof novelImage>["1"] = {
								attr: ` alt="（插圖${toFullWidth(id)}）"`,
							};

							if (ret.ok && !ret.isFile)
							{
								_options.failback = ret.input;

								return novelImage(ret.returnPath, _options);
							}

							return novelImage(ret.returnPath, _options);
						}
					}
					else
					{
						console.warn(`ID( ${innerContext} )，不存在於附件表內`);
					}
				}
				else
				{
					console.warn(`偵測到 ID( ${innerContext} )，但附件表不存在`);
				}

				return null;
			},
			default({
				tagName,
				attr,
				innerContext,
				cache,
				attach,
			})
			{
				if (tagName !== 'img' && !allowedHtmlTagList.includes(tagName))
				{
					console.warn(`not support ${tagName}`, attr, innerContext)
				}

				return null
			},
		}
	}).context;

	context = ('<div>' + context
		.replace(/^[ ]*[－＝\-—\=─–]{3,}[ ]*$/mg, '<hr/>')

		//.replace(/^([－＝\-—\=─═─＝=══－\-─—◆◇]+)$/mg, '<span class="overflow-line">$1</span>')

		.replace(/\n/g, '</div>\n<div>')
		+ '</div>')

		.replace(/<div><hr\/><\/div>/g, '<hr class="linehr"/>')

		.replace(/<div>[ ]*([－＝—=─═─＝=══－\-─—～◆◇\*＊\+＊＊↣◇◆☆★■□☆◊▃\p{Punctuation}]+)[ ]*<\/div>/ug, '<div class="linegroup calibre1 overflow-line">$1</div>')

		.replace(/<div><\/div>/g, '<div class="linegroup softbreak">　 </div>')
		.replace(/<div>/g, '<div class="linegroup calibre1">')
	;

	return context;
}

export default splitTxt
