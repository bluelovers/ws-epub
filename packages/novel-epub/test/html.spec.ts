import { IInternalProcessContextOptions } from '../lib/types';
import { IOptions } from '..';
import { defaultsDeep } from 'lodash';
import { splitTxt } from '../lib/html';

test(`full-width ruby`, () =>
{

	let actual = testHtml(`＜ｒｕｂｙ＞漢＜ｒｔ＞Ｋａｎ＜／ｒｔ＞字＜ｒｔ＞ｊｉ＜／ｒｔ＞＜／ｒｕｂｙ＞`);
	let expected;

	expect(actual).toMatch(/<ruby>漢<rt>Ｋａｎ<\/rt>字<rt>ｊｉ<\/rt><\/ruby>/);

	expect(actual).toMatchSnapshot();

});

test(`epubOptions: { iconv: 'cn' }`, () =>
{

	let actual = testHtml(`＜ｒｕｂｙ＞漢＜ｒｔ＞Ｋａｎ＜／ｒｔ＞字＜ｒｔ＞ｊｉ＜／ｒｔ＞＜／ｒｕｂｙ＞`, {
		epubOptions: {
			iconv: 'cn',
		},
	});
	let expected;

	expect(actual).toMatch(/<ruby>汉<rt>Ｋａｎ<\/rt>字<rt>ｊｉ<\/rt><\/ruby>/);

	expect(actual).toMatchSnapshot();

});

function testHtml(txt: string, plusData?: Partial<Omit<IInternalProcessContextOptions, 'epubOptions'>> & {
	epubOptions: Partial<IOptions>
})
{
	plusData = defaultsDeep(plusData || {}, {
		epub: null,
		epubOptions: {
			//iconv: 'cn'
		} as IOptions,
		store: null,
		cwd: null,
		vid: null,
		attach: null,
	});

	return splitTxt(txt, plusData as IInternalProcessContextOptions);
}
