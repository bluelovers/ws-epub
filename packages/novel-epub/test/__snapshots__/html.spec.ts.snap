// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`epubOptions: { iconv: 'cn' } 1`] = `"<div class="linegroup calibre1"><ruby>汉<rt>Ｋａｎ</rt>字<rt>ｊｉ</rt></ruby></div>"`;

exports[`full-width ruby 1`] = `"<div class="linegroup calibre1"><ruby>漢<rt>Ｋａｎ</rt>字<rt>ｊｉ</rt></ruby></div>"`;
